lib/CL/pocl_binary.c

```diff
@@ -0,0 +1,615 @@
+    buffer = orig_buffer + struct_size;
+  }
+  return CL_SUCCESS;
+}
+
+/***********************************************************/
+
+static unsigned char* tar_file(char* path, size_t basedir_offset, unsigned char* buffer)
+{
+  char* content;
+  uint64_t fsize;
+  char* p = path + basedir_offset;
+  BUFFER_STORE_STR(p);
+  pocl_read_file(path, &content, &fsize);
+  BUFFER_STORE_STR2(content, fsize);
+  free(content);
+  return buffer;
+}
+
+static unsigned char* recursively_tar_path(char* path, size_t basedir_offset, unsigned char* buffer)
+{
+  struct stat st;
+  stat(path, &st);
+
+  if (S_ISREG(st.st_mode))
+    buffer = tar_file(path, basedir_offset, buffer);
+
+  if (S_ISDIR(st.st_mode))
+    {
+      DIR *d;
+      struct dirent *entry;
+      char subpath[POCL_FILENAME_LENGTH];
+
+      strncpy(subpath, path, POCL_FILENAME_LENGTH);
+      char* p = subpath + strlen(subpath);
+      *p++ = '/';
+      d = opendir(path);
+      while ((entry = readdir(d)))
+        {
+          if (strcmp(entry->d_name, ".")==0) continue;
+          if (strcmp(entry->d_name, "..")==0) continue;
+          strcpy(p, entry->d_name);
+          buffer = recursively_tar_path(subpath, basedir_offset, buffer);
+        }
+      closedir(d);
+    }
+
+  return buffer;
+}
+
+static unsigned char* simple_tar(cl_kernel kernel, unsigned device_i, unsigned char* buffer)
+{
+  cl_program program = kernel->program;
+  char path[POCL_FILENAME_LENGTH];
+  char basedir[POCL_FILENAME_LENGTH];
+  char* p;
+  size_t len;
+  unsigned char* orig_buffer = buffer;
+
+  pocl_cache_program_path(basedir, program, device_i);
+  size_t basedir_len = strlen(basedir);
+
+  pocl_cache_program_bc_path(path, program, device_i);
+  if (pocl_exists(path))
+    buffer = tar_file(path, basedir_len, buffer);
+
+  unsigned i;
+  for (i=0; i < program->num_kernels; i++)
+    {
+      pocl_cache_kernel_cachedir(path, program, device_i, program->default_kernels[i]);
+      buffer = recursively_tar_path(path, basedir_len, buffer);
+    }
+
+  return buffer;
+}
+
+static unsigned char* pocl_binary_serialize_kernel_to_buffer(cl_kernel kernel,
+                                                   unsigned device_i,
+                                                   unsigned char *buf)
+{
+  unsigned char *buffer = buf;
+  unsigned i;
+
+  BUFFER_STORE(0, uint64_t); // struct_size
+  BUFFER_STORE(0, uint64_t); // tar_size
+  BUFFER_STORE(0, uint32_t); // arginfo size
+  size_t namelen = strlen(kernel->name);
+  BUFFER_STORE_STR2(kernel->name, namelen);
+
+  BUFFER_STORE(kernel->num_args, uint32_t);
+  BUFFER_STORE(kernel->num_locals, uint32_t);
+
+  for (i=0; i < (kernel->num_args + kernel->num_locals); i++)
+    {
+      BUFFER_STORE(kernel->dyn_arguments[i].size, uint64_t);
+    }
+
+  unsigned char *start = buffer;
+  for (i=0; i < kernel->num_args; i++)
+    {
+      pocl_argument_info *ai = &kernel->arg_info[i];
+      BUFFER_STORE(ai->access_qualifier, cl_kernel_arg_access_qualifier);
```

Comment:
Couldn't the BUFFER_STORE do swapping as easily here? Then it would be endianness independent format and portable across hosts with different endianness.